### PR TITLE
feat: test case infrastructure - consistent mockable data sources

### PR DESCRIPTION
**Pull request by mcdurdin**
* Constructs a DBDataSource that allows us to mock the data included in a test for consistent results.
* Renames some included files to .inc.php to clarify which files should be externally referenced (namespace design yet to come).

Some very basic notes on how to setup database environment, etc. More work to do here, so please don't reject this PR on the basis of configure.ps1 / README.md 🙏

### CODE REVIEW

**Comment by darcywong00** [comment]
Is there a test case for how mssql handles SMP? 

Or is it just a matter of keeping #43 ?

**Comment by mcdurdin** [comment]
> Is there a test case for how mssql handles SMP?
> 
> Or is it just a matter of keeping #43 ?

SQL Server handles it natively. I will try and include some tests. (Note that SQL Server is UTF-16 internally so the patterns of how we consume the text are different).

File: tests/data/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Data files for tests
+
+Because this data takes a while to build, and is static, we build the database once when first running any test class, based on the state of the t_dbdatasources table (we test the langtags.json record). If this matches the test data, we don't rebuild. We also always run on the first database ($mssqldb0).
+
+This means that the local database will be using test data post-test (we don't rebuild from live data), so don't forget to run tools/db/build/build_cli.php to rebuild from live data if you need it.
+
+These data files were downloaded at approximately 2020-05-25 3:40pm AEST.
+Note: langtags.json - version 1.1.1 downloaded from https://raw.githubusercontent.com/silnrsi/langtags/master/pub/langtags.json (1.1.1 is still in staging)
```
Comment (darcywong00) [inline tests/data/README.md:8]: Are the .tab files still downloaded from ethnologue? e.g.
https://www.ethnologue.com/sites/default/files/CountryCodes.tab

Comment (darcywong00) [reply tests/data/README.md:8]: nvm, I was checking if the source for the .tab files was written somewhere, but I see it in datasources.inc.php

File: tools/db/build/common.inc.php [hunk 2]
```diff
@@ -7,7 +9,7 @@
   function cache($url, $filename, $duration = 60 * 60 * 24 * 7, $do_force = false) {
```
Comment (darcywong00) [inline tools/db/build/common.inc.php:9]: $duration looks like 1 week (vs the comment "older than two weeks")

**Comment by mcdurdin** [comment]
> Is there any actual example of mocking in this PR that you could point to? I was having trouble finding one.
> 
> I presume the potential for mocking is due to the mutable `public` fields of the `DBDataSources` global? 

The mockability is in the creation of the database: we can build the database from live data or from mock data (which at this point is just a stable version of the live data) in tests/data. See [`final class TestDBDataSources extends \DBDataSources`](https://github.com/keymanapp/api.keyman.com/pull/53/files#diff-2092afd8ff10b4b675c6083a30ea1fadR18). So, for clarity, we're not mocking functions, but data.

> If so, will unit tests be responsible for unmocking once complete, or how would that work?

Once the tests are complete, the database would need to be rebuilt with live data if you wanted that. Given the tests are run not on the live site, but on a CI agent, or a developer machine, this is not a problem. I don't anticipate running the tests on the live site.

File: tools/db/build/common.inc.php (hunk 2 above)
Comment (mcdurdin) [reply tools/db/build/common.inc.php:9]: hmm, that's an old comment then :)

**Pull request merged**